src/Billboard/Billboard.sol

```diff
@@ -90,81 +107,256 @@ contract Billboard is IBillboard {
     }
 
     /// @inheritdoc IBillboard
-    function setBoardName(uint256 tokenId_, string memory name_) external {
-        registry.setBoardName(tokenId_, name_, msg.sender);
+    function setBoardName(uint256 tokenId_, string calldata name_) external isFromBoardCreator(tokenId_) {
+        registry.setBoardName(tokenId_, name_);
     }
 
     /// @inheritdoc IBillboard
-    function setBoardDescription(uint256 tokenId_, string memory description_) external {
-        registry.setBoardDescription(tokenId_, description_, msg.sender);
+    function setBoardDescription(uint256 tokenId_, string calldata description_) external isFromBoardCreator(tokenId_) {
+        registry.setBoardDescription(tokenId_, description_);
     }
 
     /// @inheritdoc IBillboard
-    function setBoardLocation(uint256 tokenId_, string memory location_) external {
-        registry.setBoardLocation(tokenId_, location_, msg.sender);
+    function setBoardLocation(uint256 tokenId_, string calldata location_) external isFromBoardCreator(tokenId_) {
+        registry.setBoardLocation(tokenId_, location_);
     }
 
     /// @inheritdoc IBillboard
-    function setBoardContentURI(uint256 tokenId_, string memory uri_) external {
-        registry.setBoardContentURI(tokenId_, uri_, msg.sender);
+    function setBoardContentURI(uint256 tokenId_, string calldata contentURI_) external isFromBoardTenant(tokenId_) {
+        registry.setBoardContentURI(tokenId_, contentURI_);
     }
 
     /// @inheritdoc IBillboard
-    function setBoardRedirectURI(uint256 tokenId_, string memory redirectURI_) external {
-        registry.setBoardRedirectURI(tokenId_, redirectURI_, msg.sender);
+    function setBoardRedirectURI(uint256 tokenId_, string calldata redirectURI_) external isFromBoardTenant(tokenId_) {
+        registry.setBoardRedirectURI(tokenId_, redirectURI_);
     }
 
     //////////////////////////////
     /// Auction
     //////////////////////////////
 
     /// @inheritdoc IBillboard
-    function setTaxRate(uint256 taxRate_) external isAdmin(msg.sender) {
-        auction.setTaxRate(taxRate_, msg.sender);
+    function getAuction(
+        uint256 tokenId_,
+        uint256 auctionId_
+    ) external view returns (IBillboardRegistry.Auction memory auction) {
+        return registry.getAuction(tokenId_, auctionId_);
     }
 
     /// @inheritdoc IBillboard
-    function getTaxRate() external view returns (uint256 taxRate) {
-        return auction.taxRate();
+    function clearAuction(uint256 tokenId_) public {
+        // revert if board not found
+        (address _boardCreator, , , , , ) = registry.boards(tokenId_);
+        if (_boardCreator == address(0)) revert BoardNotFound();
```

Comment:
`board` is a struct (like object in JS), so the boolean of `board` is always true.

Reply:
> Board memory board = registry.boards(tokenId_);
For destructing, it suppose to be gas efficient since `Board memory board` initialize a new variable in memory. Can make a gas diff later to confirm.

Reply:
can confirm that the gas diff is less than 2k...but the destructing way has risks on https://github.com/crytic/slither/wiki/Detector-Documentation#unused-return

Reply:
changed to `registry.getBoard`